### PR TITLE
ACC-2307-chore: clean up and optimize code and bumping up package versions

### PR DESCRIPTION
**Pull request by Nazehs**
This pull request contains improvements to the existing codebase of the SessionDecoder class.

Firstly, the use of the deprecated Buffer constructor has been replaced with the more secure `Buffer.from()`. Additionally, the code now uses the new `BigInt.asUintN()` function to ensure that the values for mostSig and leastSig are properly coerced to 64-bit unsigned integers, as required by the `uuidFrom()` function.

Furthermore, the uuidFrom() function has been refactored to use destructuring to unpack the values returned by the `msgpack.unpackMultiple()` function. The `toHexDigits() `function has also been updated to use the BigInt type instead of long, which is no longer needed.

Lastly, the code has been updated to use the more secure `RSA-SHA256` algorithm instead of `SHA256` for verifying the signature.

These changes result in a more secure and performant codebase for the SessionDecoder class.

However, the changes were not limited to the above. Node version was also bumped to use 16.x and 18.x both at runtime and CI and that also includes removing a single dependency(`long`) which is no longer needed and bumping up the package versions to use the latest versions possible.The tests were also updated and documentation was added to be more maintainable in the readme

[Jira ticket](https://financialtimes.atlassian.net/browse/ACC-2307)

*Note* Why i am doing this change? we might not have time to come back to this package considering we barely do any changes on it.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -9,15 +9,15 @@ references:
       - image: cimg/node:<< parameters.node-version >>-browsers
     parameters:
       node-version:
-        default: "16.14"
+        default: "16.2"
```
Comment (AniaMakes) [inline .circleci/config.yml:12]: [Blocking] missing 0

File: .circleci/config.yml [hunk 2]
```diff
@@ -75,10 +75,10 @@ jobs:
           command: git clone --depth 1 git@github.com:Financial-Times/next-ci-shared-helpers.git .circleci/shared-helpers
       - *restore_npm_cache
       - run:
-          name: Install npm v7.20.2 if node-version is 14.19
+          name: Install npm v8.19.4 if node-version is 16.2
```
Comment (AniaMakes) [inline .circleci/config.yml:78]: [Blocking] Missing 0 here as well (I think it should be v 16.20)

File: .circleci/config.yml [hunk 3]
```diff
@@ -75,10 +75,10 @@ jobs:
...
           command: |
-            if [[ "<< parameters.node-version >>" == "14.19" ]] ; then
-              npm install -g npm@7.20.2
+            if [[ "<< parameters.node-version >>" == "16.2" ]] ; then
```
Comment (AniaMakes) [inline .circleci/config.yml:80]: [Blocking] Another missing 0

File: README.md [hunk 4]
```diff
@@ -2,17 +2,27 @@
 
 [![CircleCI](https://circleci.com/gh/Financial-Times/session-decoder-js/tree/main.svg?style=svg)](https://circleci.com/gh/Financial-Times/session-decoder-js/tree/main)
 
-A simple lib that allows you to decode the uuid from an FT session.
+A library that allows you to decode the UUID from an FT session.
```
Comment (AniaMakes) [inline README.md:5]: [Praise] I really like that you replaced the "simple lib"

File: README.md [hunk 5]
```diff
@@ -2,17 +2,27 @@
...
 
-## Warning
+## Usage
 
-The session token holds the UUID but not any information as to the validity of the session - for example if the session
-was revoked by the FT or expired.
+To use the Session Decoder, you need to provide a public key to the constructor. The public key is used to verify the digital signature of the session token. Here is an example of how to use the Session Decoder to decode a session token:
 
-It is therefore unsuitable to use this library for authentication.
+```shell
```
Comment (AniaMakes) [inline README.md:11]: [Question] Is this a `shell` code or vanilla `js` code?

File: .circleci/config.yml [hunk 6]
```diff
@@ -164,13 +164,13 @@ workflows:
...
             parameters:
-              node-version: [ "16.14", "14.19" ]
+              node-version: [ "16.20", "18.15" ]
       - publish:
           context: npm-publish-token
           filters:
             <<: *filters_version_tag
           requires:
-            - test-v16.14
+            - test-v18.15
```
Comment (AniaMakes) [inline .circleci/config.yml:173]: [Question] Do we have a reason to publish if v18 test is successful vs v16? (This is not meant to be a difficult question, I am merely wondering if it's better to publish based on one vs the other set of tests)

File: src/session-decoder.js [hunk 7]
```diff
@@ -3,56 +3,63 @@
...
 }
-
-function getSessionFromTokenBuffer (sessionTokenBuffer) {
+// This function takes a session token buffer and unpacks its values to extract the most significant and least significant bits
+function getSessionFromTokenBuffer(sessionTokenBuffer) {
 	const unpackedValues = msgpack.unpackMultiple(sessionTokenBuffer);
-
-	const mostSignificantBits = new long.fromString(unpackedValues[0].toString());
-	const leastSignificantBits = new long.fromString(unpackedValues[1].toString());
+	const [mostSignificantBits, leastSignificantBits] = unpackedValues;
```
Comment (AniaMakes) [inline src/session-decoder.js:17]: [Question] I see that in the old code the values were `toString()`-ed; why did you decide not to do that in the new version?

File: src/session-decoder.js [hunk 8]
```diff
@@ -3,56 +3,63 @@
...
+function uuidFrom(mostSig, leastSig) {
+	// Convert the most and least significant bits to unsigned 64-bit integers using BigInt.asUintN
+	mostSig = BigInt.asUintN(64, mostSig);
+	leastSig = BigInt.asUintN(64, leastSig);
+	// Call the toHexDigits function with different bitshifts to create the different parts of the UUID
+	// Join the different parts of the UUID into a single string with hyphens
 	return [
-		toHexDigits(mostSig.shiftRight(32), 8),
-		toHexDigits(mostSig.shiftRight(16), 4),
+		toHexDigits(mostSig >> 32n, 8),
```
Comment (AniaMakes) [inline src/session-decoder.js:29]: [TIL] about right / left shift operators 🤯

File: src/session-decoder.js [hunk 9]
```diff
@@ -3,56 +3,63 @@
...
-function toHexDigits (val, digits) {
-	const hi = new long(1).shiftLeft((digits * 4));
-	return hi.or((val.and((hi - 1)))).toString(16).substring(1);
+// This function takes a value and a number of digits as input and returns a hexadecimal string with the specified number of digits
+function toHexDigits(val, digits) {
+	// Calculate the highest possible value based on the number of digits
+	const hi = 1n << BigInt(digits * 4);
+	// Convert the value to a hexadecimal string and substring it to get the required number of digits
+	// If the value is less than the highest possible value, the first digit will be a zero
+	return (hi | (val & (hi - 1n))).toString(16).substring(1);
```
Comment (AniaMakes) [inline src/session-decoder.js:43]: [Question] Is the pipe `|` meant to be "or" `||`? If it's meant to be a pipe, what does this mean in this context?

**Review by AniaMakes — COMMENTED** [review]
Few more questions

File: .circleci/config.yml (hunk 6 above)
Comment (Nazehs) [reply .circleci/config.yml:173]: {comment] I think the idea is to have a codebase that would support 2 versions meaning if the changes done are not passing in both versions then it shouldn't be released. I might not be 100% accurate with this but that's the mental picture i assume for this.

File: src/session-decoder.js (hunk 8 above)
Comment (jamesr101) [reply src/session-decoder.js:29]: Nice!

**Pull request merged**